### PR TITLE
Ensure stream messages are always ordered

### PR DESCRIPTION
**Pull request by fjetter**
Some motivation for this over here https://github.com/dask/distributed/pull/8049#issuecomment-1660422106

Essentially this breaks ordering of operations as soon as a handler is async. With this change, we'd ensure ordering within a given BatchedStream regardless of whether the handlers are sync or async.
I think this is a much easier to reason about behavior (but I have no idea where this will blow up... )

### CODE REVIEW

File: distributed/tests/test_core.py
Comment (fjetter) [inline distributed/tests/test_core.py:1423]: incredible that we didn't have test for this

File: distributed/core.py
Comment (fjetter) [inline distributed/core.py:1527]: This is an unrelated but pleasant addition to the CommPool

File: distributed/tests/test_core.py
Comment (hendrikmakait) [inline distributed/tests/test_core.py:1391]: I'd randomize/adjust the sleep here to test that async handlers remain in order not just because they were started in order.

File: distributed/tests/test_core.py
Comment (hendrikmakait) [inline distributed/tests/test_core.py:1431]: See above.

**Review by hendrikmakait — COMMENTED** [review]
I like the reduced complexity this approach introduces. However, I _can_ see this causing issues in real-world use cases because long-running handlers might delay others. Have you run benchmarks on this to see if anything pops up? 

The possible benefits warrant giving this a try. So, this looks good to me with one suggestion for improving the tests before giving ✅.

**Comment by hendrikmakait** [comment]
> However, I _can_ see this causing issues in real-world use cases because long-running handlers might delay others.

Just to highlight this: This _shouldn't_ cause issues, if anything, it should prevent them. I just wouldn't be surprised if there was some behavior that implicitly relied on async tasks being non-blocking and executed in the background.

**Comment by fjetter** [comment]
> I like the reduced complexity this approach introduces. However, I can see this causing issues in real-world use cases because long-running handlers might delay others. Have you run benchmarks on this to see if anything pops up?

No but reviewing the code base we actually don't even have any async handlers right now with the exception of [handle_request_refresh_who_has](https://github.com/dask/distributed/blob/405c011919bc7176bef8451be02578ca15931110/distributed/scheduler.py#L5653-L5683) although this one _should_ not be async...

**Comment by fjetter** [comment]
Well, there is `Worker.close` which is an async handler for the "terminate" operation. AFAICT this op is never used since the workers are more or less closed forcefully by terminating the stream connection. I don't think the connection termination is a very clean way to do this (and we're seeing tons of logs). However, I'll leave this to a follow up. If the close would actually block, there are ways to do this without blocking

**Comment by fjetter** [comment]
Ah sorry, the above was wrong. `terminate` is actually a dedicated-comm RPC call that is not being used. The `close` stream op is in fact used

**Comment by hendrikmakait** [comment]
> Ah sorry, the above was wrong. `terminate` is actually a dedicated-comm RPC call that is not being used. The `close` stream op is in fact used

Thanks for the clarification. Let's drop `terminate` in a dedicated PR then?

**Comment by fjetter** [comment]
> Thanks for the clarification. Let's drop terminate in a dedicated PR then?

Sure... doesn't really matter

**Review by hendrikmakait — APPROVED** [review]
LGTM, assuming CI isn't unhappier than usual.

**Pull request merged**
